QUARKUS-7367 - Implement Quantum-Safe TLS and mTLS Support in Quarkus (#323)
quarkus-qe/quarkus-test-plans: QUARKUS-7367.md

# QUARKUS-7367 Implement Quantum-Safe TLS and mTLS Support in Quarkus

JIRA: https://issues.redhat.com/browse/QUARKUS-7367

Implement Quantum-Safe TLS and mTLS Support in Quarkus

Upstream PRs:

> - https://github.com/quarkusio/quarkus/pull/55629/changes

Reviewer: Just a nitpick - we usually linked the PR'S "main page" not the "changes" page.

Documentation:
- https://quarkus.io/guides/tls-registry-reference/#post-quantum-cryptography-pqc

## Scope of testing
The QUARKUS-7367 scope is to add hybrid key exchange.
This enables to use the post-quantum cryptography (PQC) mechanism to be used alongside with the traditional algorithms.
For users this mean they don't need to update theirs certificates, keystores and truststores and still use PQC for establishing connection between server and client.

The Quarkus implements 3 enforcement policy:

> * Strict which enforce the PQC exchange and client requests which not offer PQC exchange are rejected

Reviewer: ```suggestion
* Strict which enforce the PQC exchange and client requests which not offer PQC exchange are rejected.
```

> * Client-negotiated, this mode allow for both (PQC and standard exchange) to be established, but the PQC one is preferred

Reviewer: ```suggestion
* Client-negotiated, this mode allow for both (PQC and standard exchange) to be established, but the PQC one is preferred.
```

> * Relaxed not enforce PQC and don't set any preferences

Reviewer: ```suggestion
* Relaxed not enforce PQC and don't set any preferences.
```

> * Relaxed not enforce PQC and don't set any preferences

Reviewer: The upstream documentation is a little vague on this one. It says PQC is "not enforced" but as I understand it, it is rather not even offered (it uses a different SSL engine, which does not support it).

Do you have a better understand how this works?

Author: Afaik it's just not offer PQC. I'll reword it in this TP

Currently, there is option to set `quarkus.tls.ssl-engine` to ether `openssl` or `jdkssl`.

> The tests implementation will cover only `openssl` variant as the `jdkssl` using implementation of provided JDK.

Reviewer: IHMO we should also test the `jdkssl` at least for negative scenarios.

Author: I'll add the test but the thing now is that it will fail to start, with reasonable error. For this I'll tests it.

The thing is when the PQC is backported to Java 25/21 it will probably start working and we will need to adapt tests. Or when we start supporting Java 29

Reviewer: Yea my goal with this was to test that quarkus can also handle failing scenarios reasonably. 

I think that if you put some comment to the tests like "this might start working properly with JDK 29 or JDK 25 with PQC backport" we can live with adapting test in the future.

> The JDK which support it at the moment is JDK 27.

Reviewer: ```suggestion
The JDK which support PQC at the moment is JDK 27.
```

There is plan to backport the PQC to older JDKs, but at the moment it's not done yet.
There is also plan to have it in Graal and Mandrel, when these changes are backported to JDK 25.
Currently, planned support for this feature is planned for JVM mode only on `Linux x86_64` and `Linux aarch64`.
Both Linux platform when using `openssl` engine need to have ssl library in version 3.5+.
In case of RHEL platforms, it's RHEL 10+.
This feature won't be covered on RHEL 9.

> There is possibility to extend the coverage when the PQC is backported to older JDK version, so the tests will use `jdkssl` instead of `openssl` on RHEL 9.

Reviewer: This line says which JDK will it (possibly) use on RHEL 9. Line above says that this feature **won't** be covered on RHEL 9 (at all?). Please clarify this.

Author: I tried to clarify in TP.
But simply this feature won't be covered as on RHEL 9 it's possible to use only `jdkssl` to make it work.
The PQC is not in any supported product Java so there is no need for coverage atm. If there will be need to coverage the `jdkssl`  it should be possible to extend the current coverage.

> Connection to databases with JDBC drivers won't be covered as JDBC drivers not using netty stack and some of them are not fully ready for PQC.

Reviewer: ```suggestion
Connection to databases with JDBC drivers won't be covered as JDBC drivers are not using netty stack and some of them are not fully ready for PQC.
```

This is also mention in QUARKUS-7367, that it will be part of phase 2. 

The connection to different services (Keycloak and others) won't be also covered at the moment as for each service there would be need to verification if the service support PQC.

> These connection using Quarkus TLS registry with specific "bucket".

Reviewer: ```suggestion
These connection are using Quarkus TLS registry with specific "bucket".
```

As the Quarkus TLS registry will be tested, we can trust that the PQC connection to services will also work, if the service support it.

## Automated test development

> The tests will be implemented to newly created module `http/http-pqc`.

Reviewer: IMHO it would better fit in `security/`. We also have tests for HTTPS here, and this is IMHO similar.

Author: Done

The tests will cover various combination of `quarkus.tls.pqc-enforcement-policy` and `quarkus.tls.key-exchange-groups` on server side and client offering different exchange group.
These tests will be cover both positive and negative scenario.
For example both server and client allow `X25519MLKEM768` so the connection is established, or the client support only `x25519` and server set `strict` PQC policy enforcement which end up in client failing to establish connection.

> 

Reviewer: I would like to have a scenario testing what will happen if server uses `strict` enforcement policy, but also uses `jdkssl`, or openssl binary is missing or is in old version.

There are a lot of ways how this can end with non-woking config and we should test that quarkus can detect it and report what it broken (or at least crash the app and not get stuck with non-working connections or something like it).

Author: I will add the one tests which will be executed on each platform and base on ssl version it will check the log for failed start/succesful request

As the implementation was for Quarkus TLS registry, the tests will cover default `quarkus.tls` and "bucket" `quarkus.tls` settings of TLS registry. 

These tests are planned for OpenShift and will use `registry.access.redhat.com/ubi10/openjdk-25:latest`.
The OpenShift coverage at the moment of writing this TP have some problem as the Quarkus throwing `PQC enforcement policy STRICT requires X25519MLKEM768 but the configured SSL engine does not support it`.
The UBI 10 image have ssl library 3.5+ so there should not be any problem.

> This issue for this will be created with need of enhancing the documentation as the same simple application which work on bare-metal not starting on OpenShift.

Reviewer: ```suggestion
An issue for this will be created with need of enhancing the documentation as the same simple application which work on bare-metal is not starting on OpenShift.
```

### Impact on TS and resources

> The tests will be executed on bare-metal in JVM.

Reviewer: Is there a plan for running this on OCP? 
I understand it is broken now, but it will (hopefully) get fixed.

Author: Yes I added it, but can't estimate what all will be executed

> The expected time increase will few minutes, but as the native is not covered, the build and restarts of Quarkus application will be fast.

Reviewer: ```suggestion
The expected time increase will be few minutes, but as the native is not covered, the build and restarts of Quarkus application will be fast.
```

# QUARKUS-7367 Implement Quantum-Safe TLS and mTLS Support in Quarkus

JIRA: https://issues.redhat.com/browse/QUARKUS-7367

> Implement Quantum-Safe TLS and mTLS Support in Quarkus

Reviewer 2: This part should mention that the scope is only for Netty based stack, not for the whole set of clients and extensions.

Author: Done

Upstream PRs:
- https://github.com/quarkusio/quarkus/pull/55629

Documentation:
- https://quarkus.io/guides/tls-registry-reference/#post-quantum-cryptography-pqc

## Scope of testing
The QUARKUS-7367 scope is to add hybrid key exchange.
This enables to use the post-quantum cryptography (PQC) mechanism to be used alongside with the traditional algorithms.

> For users this mean they don't need to update theirs certificates, keystores and truststores and still use PQC for establishing connection between server and client.

Reviewer 2: ```suggestion
This enables the use of a post-quantum cryptography (PQC) mechanism to be used alongside the traditional algorithms.
For users, this means they don't need to update their certificates, keystores, and truststores and still use PQC for establishing a connection between server and client.
```

The Quarkus implements 3 enforcement policy:
* Strict which enforce the PQC exchange and client requests which not offer PQC exchange are rejected.
* Client-negotiated, this mode allow for both (PQC and standard exchange) to be established, but the PQC one is preferred.

> * Relaxed not offer PQC only the standard exchange.

Reviewer 2: ```suggestion
Quarkus implements 3 enforcement policies:
* Strict, which enforces the PQC exchange, and client requests that do not offer PQC exchange are rejected.
* Client-negotiated, this mode allows both (PQC and standard exchange) to be established, but the PQC one is preferred.
* Relaxed, does not offer PQC, only the standard exchange.
```

> Currently, there is option to set `quarkus.tls.ssl-engine` to ether `openssl` or `jdkssl`.

Reviewer 2: ```suggestion
Currently, there is an option to set `quarkus.tls.ssl-engine` to either `openssl` or `jdkssl`.
```

The tests implementation will cover only `openssl` variant as the `jdkssl` using implementation of provided JDK.
The JDK which support PQC at the moment is JDK 27.
There is plan to backport the PQC to older JDKs, but at the moment it's not done yet.

> There is also plan to have it in Graal and Mandrel, when these changes are backported to JDK 25.

Reviewer 2: ```suggestion
The JDK which support PQC at the moment is JDK 27.
There is a plan to backport the PQC to older JDKs, but at the moment it's not done yet.
There is also a plan to have it in Graal and Mandrel when these changes are backported to JDK 25.
```

Currently, planned support for this feature is planned for JVM mode only on `Linux x86_64` and `Linux aarch64`.
Both Linux platform when using `openssl` engine need to have ssl library in version 3.5+.

> In case of RHEL platforms, it's RHEL 10+.

Reviewer 2: ```suggestion
Currently, support for this feature is planned for JVM mode only on `Linux x86_64` and `Linux aarch64`.
Both Linux platforms, when using the `openssl` engine, need to have the SSL library in version 3.5+.
In the case of RHEL platforms, it's RHEL 10+.
```

> This feature won't be covered on RHEL 9 in scope of test development.

Reviewer 2: I would drop this line

Author: Done

> There is possibility to extend the coverage, if needed, when the PQC is backported to older JDK version, so the tests will use `jdkssl` instead of `openssl` on RHEL 9 (coverage from QE is not planned in current scope).

Reviewer 2: ```suggestion
There is a possibility to extend the coverage, if needed, when the PQC is backported to an older JDK version, so the tests will use `jdkssl` instead of `openssl` on RHEL 9 (coverage from QE is not planned in the current scope), this will be a topic for Quarkus 4.
```

Connection to databases with JDBC drivers won't be covered as JDBC drivers are not using netty stack and some of them are not fully ready for PQC.

> This is also mention in QUARKUS-7367, that it will be part of phase 2. 

Reviewer 2: ```suggestion
Connection to databases with JDBC drivers won't be covered, as JDBC drivers do not use the Netty stack and some of them are not fully ready for PQC.
This is also mentioned in QUARKUS-7367, that it will be part of phase 2. 
```

> The connection to different services (Keycloak and others) won't be also covered at the moment as for each service there would be need to verification if the service support PQC.

Reviewer 2: ```suggestion
The connection to different services (Keycloak and others) won't be also covered at the moment as for each service there would be a need for verification if the service supports PQC.
```

These connection are using Quarkus TLS registry with specific "bucket".

> As the Quarkus TLS registry will be tested, we can trust that the PQC connection to services will also work, if the service support it.

Reviewer 2: ```suggestion
These connections are using the Quarkus TLS registry with a specific "bucket".
As the Quarkus TLS registry will be tested, we can trust that the PQC connection to services will also work if the service supports it.
```

## Automated test development

> The tests will be implemented to newly created module `security/pqc`.

Reviewer 2: ```suggestion
The tests will be implemented in the newly created module `security/pqc`.
```

The tests will cover various combination of `quarkus.tls.pqc-enforcement-policy` and `quarkus.tls.key-exchange-groups` on server side and client offering different exchange group.
These tests will be cover both positive and negative scenario.
For example both server and client allow `X25519MLKEM768` so the connection is established, or the client support only `x25519` and server set `strict` PQC policy enforcement which end up in client failing to establish connection.

> 

Reviewer 2: ```suggestion
The tests will cover various combinations of `quarkus.tls.pqc-enforcement-policy` and `quarkus.tls.key-exchange-groups` on the server side and client offering different exchange groups.
These tests will cover both positive and negative scenarios.
For example, both server and client allow `X25519MLKEM768` so the connection is established, or the client supports only `x25519` and the server sets a `strict` PQC policy enforcement, which results in the client failing to establish the connection.
```

> As the implementation was for Quarkus TLS registry, the tests will cover default `quarkus.tls` and "bucket" `quarkus.tls` settings of TLS registry. 

Reviewer 2: ```suggestion
As the implementation was for Quarkus TLS registry, the tests will cover the default `quarkus.tls` and "bucket" `quarkus.tls` settings of the TLS registry. 
```

> There will be one test to verify propper logging/failing to start application when ssl library is lower version then 3.5.

Reviewer 2: ```suggestion
There will be one test to verify proper logging/failing to start the application when the SSL library has a version lower than 3.5.
```

> 

Reviewer 2: How do you plan to do the check? Please ad a note to this line.

Author: Done

These tests are planned for OpenShift and will use `registry.access.redhat.com/ubi10/openjdk-25:latest`.
The OpenShift coverage at the moment of writing this TP have some problem as the Quarkus throwing `PQC enforcement policy STRICT requires X25519MLKEM768 but the configured SSL engine does not support it`.
The UBI 10 image have ssl library 3.5+ so there should not be any problem.
An issue for this will be created with need of enhancing the documentation as the same simple application which work on bare-metal is not starting on OpenShift.

> 

Reviewer 2: ```suggestion
The OpenShift coverage at the moment of writing this TP has some problems, as Quarkus is throwing `PQC enforcement policy STRICT requires X25519MLKEM768 but the configured SSL engine does not support it`.
The UBI 10 image has SSL library 3.5+, so there should not be any problem.
An issue for this will be created with the need to enhance the documentation, as the same simple application which work on bare-metal is not starting on OpenShift.
```

Reviewer 2: Also this looks like a blocker for OCP use-case ;/ Ping Clement about that once the issue is created

### Impact on TS and resources
The tests will be executed on bare-metal in JVM.

> The expected time increase will few minutes, but as the native is not covered, the build and restarts of Quarkus application will be fast.

Reviewer: ```suggestion
The expected time increase will be few minutes, but as the native is not covered, the build and restarts of Quarkus application will be fast.
```

Reviewer 2: Also this is a bit vague. Please quantify. Is it 2-3 minutes or more likely 8-10 or even more?

> The tests will also run on OpenShift, as there is issue on OpenShift it isn't possible to estimate impact on TS and resources.

Reviewer 2: From previous experience we know that OCP test take longer than the JVM ones. You would also need to use custom build strategy to ensure UBI 10 is used etc. 

Based on these fact you could provide some rough estimates.

Also is our testing infrastructure enabled to use custom base UBI image for the deployments?

Author: The tests will overwrite quarkus.s2i.base-jvm-image property. We did the same with UBI 8.

I update the time with rough estimate

Both Linux platforms, when using the `openssl` engine, need to have SSL library in version 3.5+.
In case the of RHEL platforms, it's RHEL 10+.
There is a possibility to extend the coverage, if needed, when the PQC is backported to an older JDK version, so the tests will use `jdkssl` instead of `openssl` on RHEL 9 (coverage from QE is not planned in the current scope), this will be a topic for Quarkus 4.

Connection to databases with JDBC drivers won't be covered, as JDBC drivers do not use Netty stack and some of them are not fully ready for PQC.
This is also mentioned in QUARKUS-7367, that it will be part of phase 2.

The connection to different services (Keycloak and others) won't be also covered at the moment as for each service there would be a need for verification if the service supports PQC.
These connections are using the Quarkus TLS registry with a specific "bucket".
As the Quarkus TLS registry will be tested, we can trust that the PQC connection to services will also work, if the service supports it.

## Automated test development

Upstream coverage:
* The upstream PR adding multiple unit tests which cover basic functionality of this feature.

> * They are no integration tests.

Reviewer 2: ```suggestion
* The upstream PR adds multiple unit tests that cover basic functionality of this feature.
* There are no integration tests.
```
